File: tests/integration/web/WorkbookFlowTest.ts
import { assert } from 'assert';
import { WebDriver, By, until } from 'selenium-webdriver';
import { WorkbookService } from '../../../src/core/services/WorkbookService';
import { AuthService } from '../../../src/core/services/AuthService';

class WorkbookFlowTest {
    private driver: WebDriver;
    private workbookService: WorkbookService;
    private authService: AuthService;

    async setUp(): Promise<void> {
        // Initialize the WebDriver
        this.driver = new WebDriver(/* configuration */);

        // Set up the WorkbookService and AuthService
        this.workbookService = new WorkbookService();
        this.authService = new AuthService();

        // Log in a test user
        await this.authService.login('testuser@example.com', 'password');
    }

    async tearDown(): Promise<void> {
        // Log out the test user
        await this.authService.logout();

        // Quit the WebDriver
        await this.driver.quit();

        // Clean up any created test data
        await this.workbookService.deleteAllTestWorkbooks();
    }

    async testCreateNewWorkbook(): Promise<void> {
        // Navigate to the home page
        await this.driver.get('https://excel.example.com');

        // Click on the 'New Workbook' button
        const newWorkbookButton = await this.driver.findElement(By.id('new-workbook-button'));
        await newWorkbookButton.click();

        // Wait for the workbook to load
        await this.driver.wait(until.elementLocated(By.id('workbook-container')), 5000);

        // Assert that a new workbook is created with default sheets
        const sheetTabs = await this.driver.findElements(By.className('sheet-tab'));
        assert.strictEqual(sheetTabs.length, 1, 'New workbook should have one default sheet');

        // Verify that the workbook title is editable
        const workbookTitle = await this.driver.findElement(By.id('workbook-title'));
        assert.strictEqual(await workbookTitle.isEnabled(), true, 'Workbook title should be editable');
    }

    async testEditCellValues(): Promise<void> {
        // Create a new workbook
        await this.testCreateNewWorkbook();

        // Enter values in multiple cells
        const cellA1 = await this.driver.findElement(By.id('cell-A1'));
        await cellA1.sendKeys('10');
        const cellB1 = await this.driver.findElement(By.id('cell-B1'));
        await cellB1.sendKeys('20');

        // Verify that the entered values are displayed correctly
        assert.strictEqual(await cellA1.getAttribute('value'), '10', 'Cell A1 should contain 10');
        assert.strictEqual(await cellB1.getAttribute('value'), '20', 'Cell B1 should contain 20');

        // Edit an existing cell value
        await cellA1.clear();
        await cellA1.sendKeys('15');

        // Verify that the change is reflected
        assert.strictEqual(await cellA1.getAttribute('value'), '15', 'Cell A1 should be updated to 15');
    }

    async testFormulaCalculation(): Promise<void> {
        // Create a new workbook
        await this.testCreateNewWorkbook();

        // Enter values in cells
        const cellA1 = await this.driver.findElement(By.id('cell-A1'));
        await cellA1.sendKeys('10');
        const cellB1 = await this.driver.findElement(By.id('cell-B1'));
        await cellB1.sendKeys('20');

        // Enter a formula in a cell
        const cellC1 = await this.driver.findElement(By.id('cell-C1'));
        await cellC1.sendKeys('=A1+B1');

        // Verify that the formula result is calculated correctly
        assert.strictEqual(await cellC1.getAttribute('value'), '30', 'Formula result should be 30');

        // Change a value in a cell referenced by the formula
        await cellA1.clear();
        await cellA1.sendKeys('15');

        // Verify that the formula result updates automatically
        assert.strictEqual(await cellC1.getAttribute('value'), '35', 'Formula result should update to 35');
    }

    async testSaveWorkbook(): Promise<void> {
        // Create a new workbook and make some changes
        await this.testEditCellValues();

        // Click the save button
        const saveButton = await this.driver.findElement(By.id('save-button'));
        await saveButton.click();

        // Verify that the save operation completes successfully
        const saveConfirmation = await this.driver.wait(until.elementLocated(By.id('save-confirmation')), 5000);
        assert.strictEqual(await saveConfirmation.isDisplayed(), true, 'Save confirmation should be displayed');

        // Reload the page
        await this.driver.navigate().refresh();

        // Verify that the changes persist after reload
        const cellA1 = await this.driver.findElement(By.id('cell-A1'));
        assert.strictEqual(await cellA1.getAttribute('value'), '15', 'Cell A1 value should persist after reload');
    }

    async testShareWorkbook(): Promise<void> {
        // Create a new workbook
        await this.testCreateNewWorkbook();

        // Click on the share button
        const shareButton = await this.driver.findElement(By.id('share-button'));
        await shareButton.click();

        // Enter another user's email and set permissions
        const shareEmailInput = await this.driver.findElement(By.id('share-email-input'));
        await shareEmailInput.sendKeys('anotheruser@example.com');
        const permissionSelect = await this.driver.findElement(By.id('permission-select'));
        await permissionSelect.sendKeys('Edit');

        // Verify that the share operation completes successfully
        const shareConfirmButton = await this.driver.findElement(By.id('share-confirm-button'));
        await shareConfirmButton.click();
        const shareConfirmation = await this.driver.wait(until.elementLocated(By.id('share-confirmation')), 5000);
        assert.strictEqual(await shareConfirmation.isDisplayed(), true, 'Share confirmation should be displayed');

        // Log in as the other user
        await this.authService.logout();
        await this.authService.login('anotheruser@example.com', 'password');

        // Verify that the shared workbook is accessible
        await this.driver.get('https://excel.example.com');
        const sharedWorkbook = await this.driver.findElement(By.xpath(`//div[contains(text(), 'Shared Workbook')]`));
        assert.strictEqual(await sharedWorkbook.isDisplayed(), true, 'Shared workbook should be accessible');
    }
}

File: tests/integration/desktop/WorkbookFlowTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Windows.Automation;
using TestStack.White;
using TestStack.White.UIItems;
using TestStack.White.UIItems.WindowItems;
using TestStack.White.Factory;
using TestStack.White.UIItems.Finders;

[TestClass]
public class WorkbookFlowTest
{
    private Application application;
    private Window mainWindow;
    private WorkbookService workbookService;
    private AuthService authService;

    [TestInitialize]
    public void TestInitialize()
    {
        // Launch the Excel desktop application
        application = Application.Launch("Excel.exe");
        
        // Initialize the main application window
        mainWindow = application.GetWindow("Microsoft Excel", InitializeOption.NoCache);
        
        // Set up the WorkbookService and AuthService
        workbookService = new WorkbookService();
        authService = new AuthService();
        
        // Log in a test user
        authService.Login("testuser@example.com", "password");
    }

    [TestCleanup]
    public void TestCleanup()
    {
        // Log out the test user
        authService.Logout();
        
        // Close the Excel application
        application.Close();
        
        // Clean up any created test data
        workbookService.CleanupTestData();
    }

    [TestMethod]
    public void TestCreateNewWorkbook()
    {
        // Click on the 'New Workbook' button in the application
        var newWorkbookButton = mainWindow.Get<Button>(SearchCriteria.ByAutomationId("NewWorkbookButton"));
        newWorkbookButton.Click();

        // Wait for the new workbook window to open
        var workbookWindow = mainWindow.Get<Window>(SearchCriteria.ByAutomationId("WorkbookWindow"));
        Assert.IsNotNull(workbookWindow, "New workbook window did not open");

        // Assert that a new workbook is created with default sheets
        var sheetTabs = workbookWindow.Get<TabStrip>(SearchCriteria.ByAutomationId("SheetTabs"));
        Assert.IsTrue(sheetTabs.TabCount > 0, "No default sheets were created");

        // Verify that the workbook title is editable
        var titleTextBox = workbookWindow.Get<TextBox>(SearchCriteria.ByAutomationId("WorkbookTitleTextBox"));
        Assert.IsTrue(titleTextBox.Enabled, "Workbook title is not editable");
    }

    [TestMethod]
    public void TestEditCellValues()
    {
        // Create a new workbook
        var newWorkbookButton = mainWindow.Get<Button>(SearchCriteria.ByAutomationId("NewWorkbookButton"));
        newWorkbookButton.Click();
        var workbookWindow = mainWindow.Get<Window>(SearchCriteria.ByAutomationId("WorkbookWindow"));

        // Enter values in multiple cells using the UI
        var cell1 = workbookWindow.Get<TextBox>(SearchCriteria.ByAutomationId("A1"));
        cell1.Enter("Test Value 1");
        var cell2 = workbookWindow.Get<TextBox>(SearchCriteria.ByAutomationId("B1"));
        cell2.Enter("Test Value 2");

        // Verify that the entered values are displayed correctly
        Assert.AreEqual("Test Value 1", cell1.Text, "Cell A1 value is incorrect");
        Assert.AreEqual("Test Value 2", cell2.Text, "Cell B1 value is incorrect");

        // Edit an existing cell value
        cell1.Enter("Updated Value");

        // Verify that the change is reflected in the UI
        Assert.AreEqual("Updated Value", cell1.Text, "Cell A1 value was not updated");
    }

    [TestMethod]
    public void TestFormulaCalculation()
    {
        // Create a new workbook
        var newWorkbookButton = mainWindow.Get<Button>(SearchCriteria.ByAutomationId("NewWorkbookButton"));
        newWorkbookButton.Click();
        var workbookWindow = mainWindow.Get<Window>(SearchCriteria.ByAutomationId("WorkbookWindow"));

        // Enter values in cells using the UI
        var cell1 = workbookWindow.Get<TextBox>(SearchCriteria.ByAutomationId("A1"));
        cell1.Enter("10");
        var cell2 = workbookWindow.Get<TextBox>(SearchCriteria.ByAutomationId("A2"));
        cell2.Enter("20");

        // Enter a formula in a cell
        var formulaCell = workbookWindow.Get<TextBox>(SearchCriteria.ByAutomationId("A3"));
        formulaCell.Enter("=SUM(A1:A2)");

        // Verify that the formula result is calculated correctly
        Assert.AreEqual("30", formulaCell.Text, "Formula result is incorrect");

        // Change a value in a cell referenced by the formula
        cell1.Enter("15");

        // Verify that the formula result updates automatically in the UI
        Assert.AreEqual("35", formulaCell.Text, "Formula result did not update automatically");
    }

    [TestMethod]
    public void TestSaveWorkbook()
    {
        // Create a new workbook and make some changes
        var newWorkbookButton = mainWindow.Get<Button>(SearchCriteria.ByAutomationId("NewWorkbookButton"));
        newWorkbookButton.Click();
        var workbookWindow = mainWindow.Get<Window>(SearchCriteria.ByAutomationId("WorkbookWindow"));
        var cell = workbookWindow.Get<TextBox>(SearchCriteria.ByAutomationId("A1"));
        cell.Enter("Test Save");

        // Click the save button in the UI
        var saveButton = workbookWindow.Get<Button>(SearchCriteria.ByAutomationId("SaveButton"));
        saveButton.Click();

        // Verify that the save operation completes successfully
        // (This might involve checking for a success message or the absence of error dialogs)

        // Close and reopen the workbook
        workbookWindow.Close();
        var openButton = mainWindow.Get<Button>(SearchCriteria.ByAutomationId("OpenButton"));
        openButton.Click();
        // (Implement logic to select the saved file in the Open dialog)

        // Verify that the changes persist after reopening
        workbookWindow = mainWindow.Get<Window>(SearchCriteria.ByAutomationId("WorkbookWindow"));
        cell = workbookWindow.Get<TextBox>(SearchCriteria.ByAutomationId("A1"));
        Assert.AreEqual("Test Save", cell.Text, "Saved changes did not persist");
    }

    [TestMethod]
    public void TestShareWorkbook()
    {
        // Create a new workbook
        var newWorkbookButton = mainWindow.Get<Button>(SearchCriteria.ByAutomationId("NewWorkbookButton"));
        newWorkbookButton.Click();
        var workbookWindow = mainWindow.Get<Window>(SearchCriteria.ByAutomationId("WorkbookWindow"));

        // Click on the share button in the UI
        var shareButton = workbookWindow.Get<Button>(SearchCriteria.ByAutomationId("ShareButton"));
        shareButton.Click();

        // Enter another user's email and set permissions
        var shareDialog = workbookWindow.Get<Window>(SearchCriteria.ByAutomationId("ShareDialog"));
        var emailInput = shareDialog.Get<TextBox>(SearchCriteria.ByAutomationId("EmailInput"));
        emailInput.Enter("anotheruser@example.com");
        var permissionComboBox = shareDialog.Get<ComboBox>(SearchCriteria.ByAutomationId("PermissionComboBox"));
        permissionComboBox.Select("Edit");
        var confirmButton = shareDialog.Get<Button>(SearchCriteria.ByAutomationId("ConfirmButton"));
        confirmButton.Click();

        // Verify that the share operation completes successfully
        // (This might involve checking for a success message or the absence of error dialogs)

        // Log in as the other user
        authService.Logout();
        authService.Login("anotheruser@example.com", "password");

        // Verify that the shared workbook is accessible
        var sharedWorkbooks = workbookService.GetSharedWorkbooks();
        Assert.IsTrue(sharedWorkbooks.Any(w => w.Name == "Shared Workbook"), "Shared workbook is not accessible");
    }
}